ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/carousel/CarouselComponentView.kt

```diff
@@ -40,16 +73,146 @@ internal fun CarouselComponentView(
+        if (carouselState.loop) {
+            Int.MAX_VALUE
+        } else {
+            pageCount
+        }
+    }
+
+    carouselState.autoAdvance?.let { autoAdvance ->
+        LaunchedEffect(Unit) {
+            while (true) {
+                delay(autoAdvance.msTimePerSlide.toLong())
+                val nextPage = if (carouselState.loop) {
+                    pagerState.currentPage + 1
+                } else {
+                    (pagerState.currentPage + 1) % pageCount
+                }
+                pagerState.animateScrollToPage(
+                    page = nextPage,
+                    animationSpec = tween(
+                        autoAdvance.msTransitionTime,
+                    ),
+                )
+            }
+        }
+    }
+
+    Column(
+        modifier = modifier
+            .padding(carouselState.margin)
+            .applyIfNotNull(shadowStyle) { shadow(it, composeShape) }
+            .applyIfNotNull(backgroundColorStyle) { background(it, composeShape) }
+            .clip(composeShape)
+            .applyIfNotNull(borderStyle) {
+                border(it, composeShape)
+                    .padding(it.width)
+            }
+            .padding(carouselState.padding),
+    ) {
+        HorizontalPager(
+            state = pagerState,
+            contentPadding = PaddingValues(horizontal = carouselState.sidePagePeek),
+            beyondViewportPageCount = pageCount,
+            pageSpacing = carouselState.spacing,
+            verticalAlignment = carouselState.alignment,
+            userScrollEnabled = style.autoAdvance == null,
+        ) { page ->
+            StackComponentView(
+                style = carouselState.slides[page % pageCount],
+                state = state,
+                clickHandler = clickHandler,
+            )
+        }
+
+        carouselState.pageControl?.let {
+            val currentPage = pagerState.currentPage % pageCount
+            PagerIndicator(
+                pageControl = it,
+                pageCount = pageCount,
+                currentPageIndex = currentPage,
+            )
+        }
+    }
+}
+
+@Composable
+private fun ColumnScope.PagerIndicator(
+    pageControl: CarouselComponentStyle.PageControlStyles,
+    pageCount: Int,
+    currentPageIndex: Int,
+    modifier: Modifier = Modifier,
+) {
+    Row(
+        modifier = modifier.align(Alignment.CenterHorizontally),
+        horizontalArrangement = Arrangement.Center,
+        verticalAlignment = Alignment.CenterVertically,
+    ) {
+        repeat(pageCount) { iteration ->
+            Indicator(if (currentPageIndex == iteration) pageControl.active else pageControl.default)
+        }
+    }
+}
+
+@Composable
+private fun Indicator(indicator: CarouselComponentStyle.IndicatorStyles) {
+    Box(
+        modifier = Modifier
+            .padding(indicator.margin)
+            .padding(horizontal = indicator.spacing / 2)
+            .clip(Shape.Pill.toShape())
```

Comment:
Micro optimization, as we don't need to use RC shapes here: 
```suggestion
            .clip(CircleShape)
```

Reply:
Right, I thought about this, but I wanted to use a pill shape here, in case someone wanted to have pill shaped indicator. Note that in the designs, width and height of the indicators are configurable, both for the active and default state.

Reply:
I added some changes to the preview so it's clear it might not have a circle shape: https://github.com/RevenueCat/purchases-android/pull/2095/commits/0012886ec677f7366a3b9205a3a8002387d1e15c